### PR TITLE
Initial go-provider integration

### PR DESCRIPTION
**Pull request by blampe**
_NB: This the first PR on top of the buildkit feature branch, which I've just cut from master._

This extends the hybrid provider to also include a native buildx provider implemented with [pulumi-go-provider](https://github.com/pulumi/pulumi-go-provider).

The image resource doesn't do anything yet, but tests are included for 100% coverage. I'd like to keep that number as high as possible.

Fixes https://github.com/pulumi/pulumi-docker/issues/884.

### CODE REVIEW

**Review by iwahbe — APPROVED** [review]
LGTM

**Review by guineveresaenger — COMMENTED** [review]
My main concerns are around the practicality of handling Yet Another Provider here. Is there a plan for reconciliation once we deem this resource production ready?

of course it is unlikely that users will use all three providers simultaneously, except perhaps as apart of an migration/upgrade.

**Comment by blampe** [comment]
Current dependencies on/for this PR:
* `feature/buildkit`
  * **PR #890** <a href="https://app.graphite.dev/github/pr/pulumi/pulumi-docker/890?utm_source=stack-comment-icon" target="_blank">Graphite</a>  👈

This [stack of pull requests](https://stacking.dev/?utm_source=stack-comment) is managed by [Graphite](https://app.graphite.dev/github/pr/pulumi/pulumi-docker/890?utm_source=stack-comment).

File: provider/hybrid.go [hunk 1]
```diff
@@ -27,19 +27,31 @@ type dockerHybridProvider struct {
 	version         string
 	bridgedProvider rpc.ResourceProviderServer
 	nativeProvider  rpc.ResourceProviderServer
+	buildxProvider  rpc.ResourceProviderServer
```
[the remark this answers is not in the document]
Comment (blampe) [reply provider/hybrid.go:30]: It's strictly for dogfooding go-pulumi-provider, which I'm using to help speed up development. Working at a higher level than `resource.PropertyMap` is a _huge_ time saver, and the integration test helpers are also great.

File: provider/resources.go [hunk 2]
```diff
@@ -217,16 +216,79 @@ func Provider() tfbridge.ProviderInfo {
...
-					{Name: "BuilderBuildKit", Value: "BuilderBuildKit",
+					{
+						Name: "BuilderBuildKit", Value: "BuilderBuildKit",
 						Description: "The builder based on moby/buildkit project",
 					},
 				},
 			},
 		},
 		ExtraResources: map[string]schema.ResourceSpec{
+			dockerResource("buildx", "Image").String(): {
```
[the remark this answers is not in the document]
Comment (blampe) [reply provider/resources.go]: Do we have the ability to alias things if this needs to move around? Is there an alternative name you'd suggest?

**Review by blampe — COMMENTED** [review]
> My main concerns are around the practicality of handling Yet Another Provider here. Is there a plan for reconciliation once we deem this resource production ready?

I think "provider" is a bit overloaded here. From the user's perspective, a provider is just a gRPC server.

From our perspective as maintainers, the "providers" here are effectively just gRPC handlers, and we happen to have 3 ways of handling requests. There's nothing wrong with that, and I'd want to understand the value we'd get from reconciling before considering it.    

> of course it is unlikely that users will use all three providers simultaneously, except perhaps as apart of an migration/upgrade.

There shouldn't be any issue mixing/matching the provider implementation, and we should expect users to at least have `Image` resources coexisting with `buildx.Image`. I'll add a test around this as part of the auth work.

File: provider/resources.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (guineveresaenger) [reply provider/resources.go]: We do have the ability to alias via the bridge; it's what we did for Imagev3 => v4 and I do believe it supports moving resources between modules as well. I was talking more from a rollout strategy point.

**Pull request closed without merging**